### PR TITLE
[CMLIB] Registry transactional writes.

### PR DESCRIPTION
**Pull request by mrmks04**
## Purpose

Implemented restore registry from log file

JIRA issue: [CORE-6762](https://jira.reactos.org/browse/CORE-6762)

## Proposed changes

- Changed create flags for log files
- Changed log write function

### CODE REVIEW

**Comment by HBelusca** [comment]
My general comment for now is: please fix the code formatting so that it sticks to the one used in the existing files (currently there are inconsistencies in: the indentation -- of parameters, etc. -- and the style of comments -- within the functions, it should be `/* xxx */` ).

Question: Is this log format the one compatible with Windows? (which is also described in https://web.archive.org/web/20201109000715/http://amnesia.gtisc.gatech.edu/~moyix/suzibandit.ltd.uk/MSc/ )

**Comment by mrmks04** [comment]
> 
> Question: Is this log format the one compatible with Windows? (which is also described in https://web.archive.org/web/20201109000715/http://amnesia.gtisc.gatech.edu/~moyix/suzibandit.ltd.uk/MSc/ )

I used this information about log files https://github.com/msuhanov/regf/blob/master/Windows%20registry%20file%20format%20specification.md#format-of-transaction-log-files
And several log dumps from Windows server 2003 analyzed

**Comment by HBelusca** [comment]
> > Question: Is this log format the one compatible with Windows? (which is also described in https://web.archive.org/web/20201109000715/http://amnesia.gtisc.gatech.edu/~moyix/suzibandit.ltd.uk/MSc/ )
> 
> I used this information about log files https://github.com/msuhanov/regf/blob/master/Windows%20registry%20file%20format%20specification.md#format-of-transaction-log-files
> And several log dumps from Windows server 2003 analyzed

OK; it would be nice to know whether all of this also confirms what has been found independently in this master thesis I linked above, in order to be sure we have a good understanding on what's going on.

File: sdk/lib/cmlib/hiveinit.c [hunk 1]
```diff
@@ -403,7 +542,16 @@ HvpGetHiveHeader(IN PHHIVE Hive,
     if (!Result) return NotHive;
 
     /* Do validation */
-    if (!HvpVerifyHiveHeader(BaseBlock)) return NotHive;
+    if (!HvpVerifyHiveHeader(BaseBlock, HFILE_TYPE_PRIMARY))
+    {
+        DPRINT1("HvpVerifyHiveHeader() failed. Try restore.\n");
+        if (!HvRestoreFromLog(Hive, BaseBlock))
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:548]: This should be done by the **callers** of `HvpGetHiveHeader()`, in case it has returned the `RecoverHeader` error and the current hive has an attached log (`Hive->Log` is TRUE).

File: sdk/lib/cmlib/hiveinit.c [hunk 2]
```diff
@@ -345,8 +476,16 @@ HvpInitializeFlatHive(
     PHHIVE Hive,
     PHBASE_BLOCK ChunkBase)
 {
-    if (!HvpVerifyHiveHeader(ChunkBase))
+    if (!HvpVerifyHiveHeader(ChunkBase, HFILE_TYPE_PRIMARY))
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:479]: Technically, flat hives don't need the log restoration, contrary to the in-memory hives (`HINIT_MEMORY`, loaded by `HvpInitializeMemoryHive()`).

File: sdk/lib/cmlib/hiveinit.c [hunk 3]
```diff
@@ -231,16 +355,23 @@ HvpInitializeMemoryHive(
...
-        !HvpVerifyHiveHeader(ChunkBase))
+    if (ChunkSize < sizeof(HBASE_BLOCK))
     {
-        DPRINT1("Registry is corrupt: ChunkSize 0x%zx < sizeof(HBASE_BLOCK) 0x%zx, "
-                "or HvpVerifyHiveHeader() failed\n", ChunkSize, sizeof(HBASE_BLOCK));
+        DPRINT1("Registry is corrupt: ChunkSize 0x%zx < sizeof(HBASE_BLOCK) 0x%zx\n", ChunkSize, sizeof(HBASE_BLOCK));
         return STATUS_REGISTRY_CORRUPT;
     }
 
+    if (!HvpVerifyHiveHeader(ChunkBase, HFILE_TYPE_PRIMARY))
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:365]: (See also the comment below). On Windows, hives loaded with `HINIT_FLAT`, `HINIT_MEMORY` or `HINIT_MEMORY_INPLACE` don't perform the LOG restoration, they just do the header verification and bail out if it's invalid.
Code that loads hives with these flags manually run a fixup routine afterwards.

Only hives loaded with `HINIT_FILE` or `HINIT_MAPFILE` do the header restoration from log when loading with `HvInitialize()`

File: sdk/lib/cmlib/hiveinit.c [hunk 4]
```diff
@@ -44,6 +45,129 @@ HvpVerifyHiveHeader(
...
+    UCHAR dirtyVectBuffer[HSECTOR_SIZE];
+    UCHAR buffer[HBLOCK_SIZE];
+
+    ASSERT(sizeof(HBASE_BLOCK) >= (HSECTOR_SIZE * Hive->Cluster));
+
+    // Read log file header
+    isSuccess = Hive->FileRead(Hive,
+                            HFILE_TYPE_LOG,
+                            &offset,
+                            &logBaseBlock,
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:67]: Normally, code will allocate a block, like there:
https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/cmlib/hiveinit.c;hb=a8d2cd4b9d124bf9a9d97105cb11923827d7142b#l386
to be used in this `FileRead` call afterwards.

File: sdk/lib/cmlib/hiveinit.c [hunk 5]
```diff
@@ -44,6 +45,129 @@ HvpVerifyHiveHeader(
...
+    if (!isSuccess)
+    {
+        DPRINT1("Read LOG file failed\n");
+        return FALSE;
+    }
+
+    // Validate log header
+    if (!HvpVerifyHiveHeader(&logBaseBlock, HFILE_TYPE_LOG))
+    {
+        DPRINT1("LOG header corrupted\n");
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:79]: TODO: Log can self-heal (depending on the value of the `CmpSelfHeal` variable, and on some flags set in the `CmpBootType` variable).

File: sdk/lib/cmlib/hiveinit.c [hunk 6]
```diff
@@ -44,6 +45,129 @@ HvpVerifyHiveHeader(
...
+    // Read log file header
+    isSuccess = Hive->FileRead(Hive,
+                            HFILE_TYPE_LOG,
+                            &offset,
+                            &logBaseBlock,
+                            Hive->Cluster * HSECTOR_SIZE);
+
+    if (!isSuccess)
+    {
+        DPRINT1("Read LOG file failed\n");
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:72]: Self-healing possible? (see below)

File: sdk/lib/cmlib/hiveinit.c [hunk 7]
```diff
@@ -44,6 +45,129 @@ HvpVerifyHiveHeader(
...
+        return FALSE;
+    }
+
+    // Validate log header
+    if (!HvpVerifyHiveHeader(&logBaseBlock, HFILE_TYPE_LOG))
+    {
+        DPRINT1("LOG header corrupted\n");
+        return FALSE;
+    }
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:81]: What has been done until now, is for restoring the header when handling the `RecoverHeader` error case during loading.
What is following next below, is for handling the `RecoverData` error case during loading.
It would be nice to have these in two separate recovery functions.

File: ntoskrnl/config/cminit.c [hunk 8]
```diff
@@ -545,14 +545,14 @@ CmpOpenHiveFiles(IN PCUNICODE_STRING BaseName,
 
     /* Now create the file */
     Status = ZwCreateFile(Log,
-                          DesiredAccess,
+                          DesiredAccess | SYNCHRONIZE,
```
Comment (HBelusca) [inline ntoskrnl/config/cminit.c:548]: Why is this needed? (same question for the IoFlags too)

File: drivers/filesystems/fastfat_new/create.c [hunk 9]
```diff
@@ -629,7 +629,7 @@ Return Value:
 
     ULONG MatchFlags = 0;
 
-    CCB LocalCcb;
+    CCB LocalCcb = {0};
```
Comment (HBelusca) [inline drivers/filesystems/fastfat_new/create.c:632]: What problem does this fix? Is it fixing an uninitialized variable usage?

File: sdk/lib/cmlib/hivewrt.c [hunk 10]
```diff
@@ -104,12 +110,13 @@ HvpWriteLog(
         FileOffset += HBLOCK_SIZE;
     }
 
-    Success = RegistryHive->FileSetSize(RegistryHive, HFILE_TYPE_LOG, FileOffset, FileOffset);
+    /* TODO: Check this function. In some cases log file can be corrrupted */
+    /* Success = RegistryHive->FileSetSize(RegistryHive, HFILE_TYPE_LOG, FileOffset, FileOffset);
```
Comment (HBelusca) [inline sdk/lib/cmlib/hivewrt.c:114]: It's preferrable to use `#if 0 ... #endif` for commenting out large blocks of code btw.

File: drivers/filesystems/fastfat_new/create.c (hunk 9 above)
Comment (mrmks04) [reply drivers/filesystems/fastfat_new/create.c:632]: Yes, uninitialized usage

File: ntoskrnl/config/cminit.c (hunk 8 above)
Comment (mrmks04) [reply ntoskrnl/config/cminit.c:548]: Without this flag, it have stack corrupt. It happen if inside RegistryHive->FileWrite function ZwWriteFile return STATUS_PENDING. Or in RegistryHive->FileFlush function ZwFlushBuffersFile return STATUS_PENDING.

File: drivers/filesystems/fastfat_new/create.c (hunk 9 above)
Comment (HBelusca) [reply drivers/filesystems/fastfat_new/create.c:632]: This may be worth going into a separate PR to be committed asap.

Comment (mrmks04) [reply drivers/filesystems/fastfat_new/create.c:632]: Ok

File: sdk/lib/cmlib/hiveinit.c [hunk 11]
```diff
@@ -441,10 +590,38 @@ HvLoadHive(IN PHHIVE Hive,
...
+            }
+            #endif 
+            
+            if (!HvpRecoverHeaderFromLog(Hive, &BaseBlock))
+            {
+                DPRINT1("Recover header from LOG failed\n");
+                return STATUS_REGISTRY_CORRUPT;
+            }
+
+            DPRINT1("Hive header recovered\n");
```
Comment (Extravert-ir) [inline sdk/lib/cmlib/hiveinit.c:615]: This DPRINT is ambiguous, the last one is enough

File: drivers/filesystems/fastfat_new/create.c (hunk 9 above)
Comment (JoachimHenze) [reply drivers/filesystems/fastfat_new/create.c:632]: Yes please commit that line via a separate commit/PR!

Comment (Extravert-ir) [reply drivers/filesystems/fastfat_new/create.c:632]: Is it still the case? Seems like nobody else spotted this, maybe that was MSVC RTC bug we have hacked around recently?

Comment (mrmks04) [reply drivers/filesystems/fastfat_new/create.c:632]: Here my PR https://github.com/reactos/reactos/pull/3934

Comment (JoachimHenze) [reply drivers/filesystems/fastfat_new/create.c:632]: That PR has been closed and superseded by a more correct approach by Timo. Can you remove that single-line change now here? Is the problem gone now that you intended to address with that line?

File: ntoskrnl/config/cminit.c (hunk 8 above)
Comment (JoachimHenze) [reply ntoskrnl/config/cminit.c:548]: Does that claim still stand @mrmks04 with todays master head? Again see the already merged https://github.com/reactos/reactos/pull/3679

File: drivers/filesystems/fastfat_new/create.c (hunk 9 above)
Comment (mrmks04) [reply drivers/filesystems/fastfat_new/create.c:632]: PR https://github.com/reactos/reactos/pull/3909 fix my problem

File: ntoskrnl/config/cminit.c (hunk 8 above)
Comment (mrmks04) [reply ntoskrnl/config/cminit.c:548]: With latest code from master problem gone.

**Comment by HBelusca** [comment]
PR is in review.

File: sdk/lib/cmlib/hiveinit.c [hunk 12]
```diff
@@ -362,6 +363,150 @@ HvpInitializeFlatHive(
     return STATUS_SUCCESS;
 }
 
+/**
+ * @name HvpRecoverHeaderFromLog
+ *
+ * Function to recover hive header from log.
+ */
+BOOLEAN CMAPI
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:371]: Suggestion: use instead a `ULONG` return value; this will allow you to specify more accurate error codes than TRUE/FALSE (see after). This would allow the caller (see `HvLoadHive`) to translate this hive-lib-specific error codes into their corresponding STATUS values.

File: sdk/lib/cmlib/hiveinit.c [hunk 13]
```diff
@@ -362,6 +363,150 @@ HvpInitializeFlatHive(
...
+    ULONG Offset = 0;
+    BOOLEAN IsSuccess;
+
+    ASSERT(sizeof(HBASE_BLOCK) >= (HSECTOR_SIZE * Hive->Cluster));
+
+    LogBaseBlock = HvpAllocBaseBlockAligned(Hive, TRUE, TAG_CM);
+    if (!LogBaseBlock)
+    {
+        DPRINT1("Allocate base block failed\n");
+        return FALSE;
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:386]: Here for example it would be `NoMemory`.

File: sdk/lib/cmlib/hiveinit.c [hunk 14]
```diff
@@ -362,6 +363,150 @@ HvpInitializeFlatHive(
...
+
+    if (!IsSuccess)
+    {
+        DPRINT1("Write hive header failed\n");
+        return FALSE;
+    }
+
+    *BaseBlock = LogBaseBlock;
+
+    return TRUE;
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:430]: Here `HiveSuccess`.

File: sdk/lib/cmlib/hiveinit.c [hunk 15]
```diff
@@ -362,6 +363,150 @@ HvpInitializeFlatHive(
...
+    IsSuccess = Hive->FileRead(Hive,
+                               HFILE_TYPE_LOG,
+                               &Offset,
+                               LogBaseBlock,
+                               Hive->Cluster * HSECTOR_SIZE);
+
+    if (!IsSuccess)
+    {
+        DPRINT1("Read LOG file failed\n");
+        return FALSE;
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:399]: Here and below, all the failures would be `Fail`.

File: sdk/lib/cmlib/hiveinit.c [hunk 16]
```diff
@@ -362,6 +363,150 @@ HvpInitializeFlatHive(
...
+
+    return TRUE;
+}
+
+/**
+ * @name HvpRecoverDataLog
+ *
+ * Function to recover hive data from log.
+ */
+BOOLEAN CMAPI
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:438]: Same suggestion here as for the previous Recover function.

File: sdk/lib/cmlib/hiveinit.c [hunk 17]
```diff
@@ -16,12 +16,13 @@
...
 HvpVerifyHiveHeader(
-    IN PHBASE_BLOCK BaseBlock)
+    IN PHBASE_BLOCK BaseBlock,
+    IN ULONG HfileType)
 {
     if (BaseBlock->Signature != HV_HBLOCK_SIGNATURE ||
         BaseBlock->Major != HSYS_MAJOR ||
         BaseBlock->Minor < HSYS_MINOR ||
-        BaseBlock->Type != HFILE_TYPE_PRIMARY ||
+        BaseBlock->Type != HfileType ||
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:25]: Since you generalized this function to be also usable for `HFILE_TYPE_LOG` types :+1: :
have you checked whether Windows registry LOG files' base block have all the Major and Minor fields set?

File: sdk/lib/cmlib/hiveinit.c [hunk 18]
```diff
@@ -362,6 +363,150 @@ HvpInitializeFlatHive(
...
+
+    if (*((PULONG)DirtyVectBuffer) != DIRTY_ID)
+    {
+        DPRINT1("Wrong header in dirty block\n");
+        return FALSE;
+    }
+
+    IndexInLog = 0;
+    /* Write birty blocks */
+    for (ULONG blockIndex = 0; blockIndex < BaseBlock->Length / HBLOCK_SIZE; ++blockIndex)
```
Comment (HBelusca) [inline sdk/lib/cmlib/hiveinit.c:471]: Keep the existing style of this hive library and move the blockIndex variable definition at the top of the function.

File: sdk/lib/cmlib/hiveinit.c [hunk 19]
```diff
@@ -441,10 +590,36 @@ HvLoadHive(IN PHHIVE Hive,
...
+            }
+            #endif 
+            
+            if (!HvpRecoverHeaderFromLog(Hive, &BaseBlock))
+            {
+                DPRINT1("Recover header from LOG failed\n");
+                return STATUS_REGISTRY_CORRUPT;
+            }
+
+            if (!HvpRecoverDataFromLog(Hive, BaseBlock))
```
[the remark this answers is not in the document]
Comment (HBelusca) [reply sdk/lib/cmlib/hiveinit.c:615]: Shouldn't the data recovery be done after the whole hive be read? (see the `/* Now read the whole hive */` below)

File: ntoskrnl/config/cminit.c [hunk 20]
```diff
@@ -545,14 +545,14 @@ CmpOpenHiveFiles(IN PCUNICODE_STRING BaseName,
...
                           &ObjectAttributes,
                           &IoStatusBlock,
                           NULL,
                           AttributeFlags,
                           ShareMode,
                           CreateDisposition,
-                          IoFlags,
+                          IoFlags | FILE_SYNCHRONOUS_IO_NONALERT,
```
[the remark this answers is not in the document]
Comment (mrmks04) [reply ntoskrnl/config/cminit.c:555]: Windows server 2003 open file with flags 0x40008, for ntuser.dat.log

File: sdk/lib/cmlib/hiveinit.c (hunk 17 above)
Comment (mrmks04) [reply sdk/lib/cmlib/hiveinit.c:25]: Base block in log file is copy of original registry base block.

Comment (HBelusca) [reply sdk/lib/cmlib/hiveinit.c:25]: OK so it's better to check for this.

File: sdk/lib/cmlib/hiveinit.c (hunk 16 above)
Comment (mrmks04) [reply sdk/lib/cmlib/hiveinit.c:438]: Not realy. "recover hive data" and "recover hive header".:smiley:

File: sdk/lib/cmlib/hivewrt.c [hunk 21]
```diff
@@ -104,12 +110,15 @@ HvpWriteLog(
         FileOffset += HBLOCK_SIZE;
     }
 
+    /* TODO: Check this function. In some cases log file can be corrrupted */
```
[the remark this answers is not in the document]
Comment (mrmks04) [reply sdk/lib/cmlib/hivewrt.c:112]: But file size calculated in while cycle before call this function. I think it neded to test with latest master.

File: sdk/lib/cmlib/hivewrt.c [hunk 22]
```diff
@@ -63,9 +55,23 @@ HvpWriteLog(
...
+    while (BlockIndex < RegistryHive->Storage[Stable].Length)
+    {
+        LastIndex = BlockIndex;
+        BlockIndex = RtlFindSetBits(&RegistryHive->DirtyVector, 1, BlockIndex);
+        if (BlockIndex == ~0U || BlockIndex < LastIndex)
+        {
+            break;
+        }
+
+        Ptr[BlockIndex] = DIRTY_BLOCK;
```
[the remark this answers is not in the document]
Comment (mrmks04) [reply sdk/lib/cmlib/hivewrt.c:71]: In ReactOS vector contains 1 bit per 1 block (4kb), but in Windows 1 bit per 1 sector (512 bytes). In Windows log file dirty vector contains only 0x00 and 0xFF values, I don't see othrers. And content write only by blocks.

[the remark this answers is not in the document]
Comment (HBelusca) [reply sdk/lib/cmlib/hivewrt.c:71]: > In ReactOS vector contains 1 bit per 1 block (4kb),

This will need to be changed at some point, for NT/Windows compatibility.

**Comment by GeoB99** [comment]
Superseded by #4571. Thank you for contributing nonetheless!
